drivers/serial/uart_stm32.c

```diff
@@ -2014,6 +2014,105 @@ static void uart_stm32_irq_config_func_##index(const struct device *dev)	\
 #define STM32_UART_PM_WAKEUP(index) /* Not used */
 #endif
 
+/* Ensure DTS doesn't present an incompatible parity configuration.
+ * Mark/space parity isn't supported on the STM32 family.
+ * If 9 data bits are configured, ensure that a parity bit isn't set.
+ */
+#define STM32_UART_CHECK_DT_PARITY(index)				\
+BUILD_ASSERT(								\
+	!(DT_INST_ENUM_IDX_OR(index, parity, STM32_UART_DEFAULT_PARITY)	\
+					== UART_CFG_PARITY_MARK ||	\
+	DT_INST_ENUM_IDX_OR(index, parity, STM32_UART_DEFAULT_PARITY)	\
+					== UART_CFG_PARITY_SPACE),	\
+	"Node " DT_NODE_PATH(DT_DRV_INST(index))			\
+	" has unsupported parity configuration");			\
+BUILD_ASSERT(								\
+	!(DT_INST_ENUM_IDX_OR(index, parity, STM32_UART_DEFAULT_PARITY)	\
+					!= UART_CFG_PARITY_NONE	&&	\
+	DT_INST_ENUM_IDX_OR(index, data_bits,				\
+			    STM32_UART_DEFAULT_DATA_BITS)		\
+					== UART_CFG_DATA_BITS_9),	\
+	"Node " DT_NODE_PATH(DT_DRV_INST(index))			\
+		" has unsupported parity + data bits combination");
+
+/* Ensure DTS doesn't present an incompatible data bits configuration
+ * The STM32 family doesn't support 5 or 6 data bits, only some series support 7
+ */
+#ifdef LL_USART_DATAWIDTH_7B
+#define STM32_UART_CHECK_DT_DATA_BITS(index)				\
+BUILD_ASSERT(								\
+	!(DT_INST_ENUM_IDX_OR(index, data_bits,				\
+			      STM32_UART_DEFAULT_DATA_BITS)		\
+					== UART_CFG_DATA_BITS_5 ||	\
+	DT_INST_ENUM_IDX_OR(index, data_bits,				\
+			    STM32_UART_DEFAULT_DATA_BITS)		\
+					== UART_CFG_DATA_BITS_6),	\
+	"Node " DT_NODE_PATH(DT_DRV_INST(index))			\
+		" has unsupported data bits configuration");
+#else
+#define STM32_UART_CHECK_DT_DATA_BITS(index)				\
+BUILD_ASSERT(								\
+	!(DT_INST_ENUM_IDX_OR(index, data_bits,				\
+			      STM32_UART_DEFAULT_DATA_BITS)		\
+					== UART_CFG_DATA_BITS_5 ||	\
+	DT_INST_ENUM_IDX_OR(index, data_bits,				\
+			    STM32_UART_DEFAULT_DATA_BITS)		\
+					== UART_CFG_DATA_BITS_6 ||	\
+	DT_INST_ENUM_IDX_OR(index, data_bits,				\
+			    STM32_UART_DEFAULT_DATA_BITS)		\
+					== UART_CFG_DATA_BITS_7),	\
```

Comment:
Likewise, 7-bit payload + parity is a valid case of LL_USART_DATAWIDTH_8B and should not spring an assert

Reply:
Ditto for my comment above.

Reply:
Adapted the check to ensure that parity is set if 7-data bits are defined with `LL_USART_DATAWIDTH_8B`.